[usbdev] Stall IN/OUT transactions of unimplemented endpoints

vogelpi opened the pull request:
This is mandated by the spec. Previously, they were ignored just like SETUP transactions for unimplemented endpoints.

eunchan commented:
Sorry Pirmin, I am not familiar with USB spec. But could you please add description of what "stall in/out transactions" means vs previous implementation (treat them as SETUP?) Based on the code, looks like (correct me if I am wrong) the code now tries to keep the output but removing the guard if inputs?

vogelpi commented:
Thanks @mdhayter and @eunchan for your feedback.

In general, stalling is used to signal that the host should not retry a particular transaction or to signal certain error conditions.

Previously we ignored all IN, OUT and SETUP transactions of unimplemented endpoints. However, the spec mandates that for unimplemented endpoints:
- IN/OUT transactions are stalled, and
- SETUP transactions are ignored.

This is what this PR does.

When studying the implementation to understand how to best change the behaviour for IN/OUT transactions of unimplemented endpoints, I noted that our module already supports a feature to stall IN/OUT transactions on a per-endpoint basis (software can switch a bit in a config register). Thus, I simply adapted the FSMs for IN/OUT transactions to behave the same if
- the specified endpoint is configured to stall IN/OUT transactions (existing feature)
- the specified endpoint is not implemented (new)

In both cases, the transaction is now stalled.

vogelpi commented:
Update: rebased on top of master, added a function to the DPI model to send SETUP, OUT and IN transactions to an unimplemented endpoint.

At the moment, the DPI model cannot react on responses from the device but we have some basic observability via `monitor_usb()` which logs all packets sent between host and device. Successful stalling of OUT, IN transactions and ignoring of SETUP transactions to unimplemented endpoints can be observed in the log file.

vogelpi commented:
Thanks @tomroberts-lowrisc for your review and approval. I am merging this now.
